File: persistence/src/main/java/de/cyface/persistence/v1/model/ParcelablePoint3D.java
package de.cyface.persistence.v1.model;

import android.os.Parcel;
import android.os.Parcelable;

/**
 * A measured {@code DataPoint} with three coordinates such as an acceleration-, rotation- or magnetic value point.
 *
 * @author Klemens Muthmann
 * @version 3.0.0
 * @since 1.0.0
 */
public final class ParcelablePoint3D extends DataPoint implements de.cyface.model.Point3D {
    /**
     * The x component of the data point.
     */
    private final float x;
    /**
     * The y component of the data point.
     */
    private final float y;
    /**
     * The z component of the data point.
     */
    private final float z;

    /**
     * Creates a new completely initialized <code>Point3D</code>.
     *
     * @param identifier The database wide unique identifier of this data point.
     * @param x The x component of the data point.
     * @param y The y component of the data point.
     * @param z The z component of the data point.
     * @param timestamp The time when this point was measured in milliseconds since 1.1.1970.
     */
    public ParcelablePoint3D(final Long identifier, final float x, final float y, final float z, final long timestamp) {
        super(identifier, timestamp);
        this.x = x;
        this.y = y;
        this.z = z;
    }

    /**
     * Creates a new completely initialized <code>Point3D</code>.
     *
     * @param x The x component of the data point.
     * @param y The y component of the data point.
     * @param z The z component of the data point.
     * @param timestamp The time when this point was measured in milliseconds since 1.1.1970.
     */
    public ParcelablePoint3D(final float x, final float y, final float z, final long timestamp) {
        this(null, x, y, z, timestamp);
    }

    /**
     * @return The x component of the data point.
     */
    public float getX() {
        return x;
    }

    /**
     * @return The y component of the data point.
     */
    public float getY() {
        return y;
    }

    /**
     * @return The z component of the data point.
     */
    public float getZ() {
        return z;
    }

    /*
     * MARK: Non Bean Code.
     */

    @Override
    public String toString() {
        return "Point X: " + x + "Y: " + y + "Z: " + z;
    }

    /*
     * MARK: Code For Parcelable Interface.
     */

    /**
     * Recreates this point from the provided <code>Parcel</code>.
     *
     * @param in Serialized form of a <code>Point3D</code>.
     */
    public ParcelablePoint3D(final Parcel in) {
        super(in);
        this.x = in.readFloat();
        this.y = in.readFloat();
        this.z = in.readFloat();
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        super.writeToParcel(dest, flags);
        dest.writeFloat(x);
        dest.writeFloat(y);
        dest.writeFloat(z);
    }

    /**
     * The <code>Parcelable</code> creator as required by the Android Parcelable specification.
     */
    public static final Parcelable.Creator<ParcelablePoint3D> CREATOR = new Parcelable.Creator<>() {

        @Override
        public ParcelablePoint3D createFromParcel(Parcel source) {
            return new ParcelablePoint3D(source);
        }

        @Override
        public ParcelablePoint3D[] newArray(int size) {
            return new ParcelablePoint3D[size];
        }
    };
}

File: persistence/src/main/java/de/cyface/persistence/model/ParcelablePoint3D.kt
/*
 * Copyright 2023 Cyface GmbH
 *
 * This file is part of the Cyface SDK for Android.
 *
 * The Cyface SDK for Android is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * The Cyface SDK for Android is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with the Cyface SDK for Android. If not, see <http://www.gnu.org/licenses/>.
 */
package de.cyface.persistence.model

import android.os.Parcel
import android.os.Parcelable.Creator
import de.cyface.model.Point3D
import java.util.Objects

/**
 * This class represents a [DataPoint] with three coordinates such as an acceleration-, rotation- or direction point.
 *
 * @author Klemens Muthmann
 * @author Armin Schnabel
 * @version 4.0.0
 * @since 1.0.0
 */
open class ParcelablePoint3D : DataPoint, Point3D {
    /**
     * The x component of the data point.
     */
    override val x: Float
    /**
     * The y component of the data point.
     */
    override val y: Float
    /**
     * The z component of the data point.
     */
    override val z: Float

    /**
     * Creates a new completely initialized instance of this class.
     *
     * @param timestamp The timestamp at which this data point was captured in milliseconds since 1.1.1970.
     * @param x The x component of the data point.
     * @param y The y component of the data point.
     * @param z The z component of the data point.
     */
    constructor(timestamp: Long, x: Float, y: Float, z: Float) : super(timestamp) {
        this.x = x
        this.y = y
        this.z = z
    }

    /*
     * MARK: Parcelable Interface
     */

    /**
     * Constructor as required by `Parcelable` implementation.
     *
     * @param in A `Parcel` that is a serialized version of a data point.
     */
    protected constructor(`in`: Parcel) : super(`in`) {
        x = `in`.readFloat()
        y = `in`.readFloat()
        z = `in`.readFloat()
    }

    override fun describeContents(): Int {
        return 0
    }

    override fun writeToParcel(dest: Parcel, flags: Int) {
        super.writeToParcel(dest, flags)
        dest.writeFloat(x)
        dest.writeFloat(y)
        dest.writeFloat(z)
    }

    override fun toString(): String {
        return "Point3d(x=$x, y=$y, z=$z)"
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || javaClass != other.javaClass) return false
        if (!super.equals(other)) return false
        val point3D1 = other as ParcelablePoint3D
        return point3D1.x.compareTo(x) == 0
    }

    override fun hashCode(): Int {
        return Objects.hash(super.hashCode(), x, y, z)
    }

    companion object {
        /**
         * The `Parcelable` creator as required by the Android Parcelable specification.
         */
        @JvmField
        val CREATOR: Creator<ParcelablePoint3D?> = object : Creator<ParcelablePoint3D?> {
            override fun createFromParcel(`in`: Parcel): ParcelablePoint3D {
                return ParcelablePoint3D(`in`)
            }

            override fun newArray(size: Int): Array<ParcelablePoint3D?> {
                return arrayOfNulls(size)
            }
        }
    }
}